File: src/HealthcareSupport.CaseEvaluation.Application/ExternalSignups/ExternalUserDisplayName.cs
namespace HealthcareSupport.CaseEvaluation.ExternalSignups;

/// <summary>
/// Phase 1 / C2 / D4 (firm-based AA/DA registration, 2026-06-11) -- shared
/// display-name resolution for external users.
///
/// <para>Firm AA/DA accounts register with a blank <c>IdentityUser.Name</c>/
/// <c>Surname</c> and only a <c>FirmName</c> extension property (D1). Without
/// a fallback, every place that renders an external user's name shows a blank
/// or the raw email. This helper resolves a single display string.</para>
///
/// <para>Precedence (locked by Adrian, Q1 2026-06-11 -- fall back to firm name
/// when first/last are absent): "First Last" if present, else FirmName, else
/// email, else empty. Whitespace-only inputs are treated as absent.</para>
///
/// <para>Pure (no DI / no DB): callers pass the already-loaded values. The TS
/// mirror is <c>resolveExternalUserDisplayName</c> in
/// <c>angular/src/app/shared/</c>; keep the two precedence rules in sync.</para>
/// </summary>
internal static class ExternalUserDisplayName
{
    internal static string Resolve(string? firstName, string? lastName, string? firmName, string? email)
    {
        var fullName = $"{firstName?.Trim()} {lastName?.Trim()}".Trim();
        if (!string.IsNullOrWhiteSpace(fullName))
        {
            return fullName;
        }

        if (!string.IsNullOrWhiteSpace(firmName))
        {
            return firmName.Trim();
        }

        return email?.Trim() ?? string.Empty;
    }
}

File: angular/src/app/shared/auth/external-user-display-name.ts
/**
 * Phase 1 / C2 / D4 (firm-based AA/DA registration, 2026-06-11) -- TS mirror of
 * the C# `ExternalUserDisplayName.Resolve` helper
 * (src/.../Application/ExternalSignups/ExternalUserDisplayName.cs).
 *
 * Firm AA/DA accounts register with a blank IdentityUser name/surname and only
 * a FirmName extension property (D1). This resolves a single display string for
 * the home banner, external-user pickers, and party tables so a firm account
 * shows its firm name instead of a blank or the raw email.
 *
 * Precedence (locked by Adrian, Q1 2026-06-11 -- fall back to firm name when
 * first/last are absent): "First Last" if present, else FirmName, else email,
 * else "". Whitespace-only inputs are treated as absent. Keep the two precedence
 * rules in sync with the C# helper.
 */
export function resolveExternalUserDisplayName(
  firstName?: string | null,
  lastName?: string | null,
  firmName?: string | null,
  email?: string | null,
): string {
  const fullName = `${(firstName ?? '').trim()} ${(lastName ?? '').trim()}`.trim();
  if (fullName) {
    return fullName;
  }

  const firm = firmName?.trim();
  if (firm) {
    return firm;
  }

  return email?.trim() ?? '';
}
